controller/util/cf_v2_client.go

```diff
@@ -0,0 +1,172 @@
+	"net/http"
+	"time"
+
+	model "github.com/kubernetes-incubator/service-catalog/model/service_broker"
+	scmodel "github.com/kubernetes-incubator/service-catalog/model/service_controller"
+)
+
+const (
+	catalogFormatString         = "%s/v2/catalog"
+	serviceInstanceFormatString = "%s/v2/service_instances/%s"
+	bindingFormatString         = "%s/v2/service_instances/%s/service_bindings/%s"
+
+	httpTimeoutSeconds = 15
+)
+
+type cfV2BrokerClient struct {
+	broker *scmodel.ServiceBroker
+	client *http.Client
+}
+
+// CreateCFV2BrokerClient creates an instance of BrokerClient for communicating
+// with brokers which implement the Cloud Foundry Service Broker V2 API.
+func CreateCFV2BrokerClient(b *scmodel.ServiceBroker) BrokerClient {
+	return &cfV2BrokerClient{
+		broker: b,
+		client: &http.Client{
+			Timeout: httpTimeoutSeconds * time.Second,
+		},
+	}
+}
+
+func (c *cfV2BrokerClient) GetCatalog() (*model.Catalog, error) {
+	url := fmt.Sprintf(catalogFormatString, c.broker.BrokerURL)
+
+	req, err := http.NewRequest("GET", url, nil)
+	req.SetBasicAuth(c.broker.AuthUsername, c.broker.AuthPassword)
+	resp, err := c.client.Do(req)
+	if err != nil {
+		log.Printf("Failed to fetch catalog from %s\n%v", url, resp)
+		log.Printf("err: %#v", err)
+		return nil, err
+	}
+
+	var catalog model.Catalog
+	if err = ResponseBodyToObject(resp, &catalog); err != nil {
+		log.Printf("Failed to unmarshal catalog: %#v", err)
+		return nil, err
+	}
+
+	return &catalog, nil
+}
+
+func (c *cfV2BrokerClient) CreateServiceInstance(ID string, req *model.ServiceInstanceRequest) (*model.ServiceInstance, error) {
+	jsonBytes, err := json.Marshal(req)
+	if err != nil {
+		return nil, err
+	}
+
+	url := fmt.Sprintf(serviceInstanceFormatString, c.broker.BrokerURL, ID)
+
+	// TODO: Handle the auth
+	createHTTPReq, err := http.NewRequest("PUT", url, bytes.NewReader(jsonBytes))
+	log.Printf("Doing a request to: %s", url)
+	resp, err := c.client.Do(createHTTPReq)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	// TODO: Align this with the actual response model.
+	si := model.ServiceInstance{}
+	if err := ResponseBodyToObject(resp, &si); err != nil {
+		return nil, err
+	}
+	return &si, nil
+}
+
+func (c *cfV2BrokerClient) UpdateServiceInstance(ID string, req *model.ServiceInstanceRequest) (*model.ServiceInstance, error) {
+	// TODO: https://github.com/kubernetes-incubator/service-catalog/issues/114
+	return nil, fmt.Errorf("Not implemented")
+}
+
+func (c *cfV2BrokerClient) DeleteServiceInstance(ID string) error {
+	url := fmt.Sprintf(serviceInstanceFormatString, c.broker.BrokerURL, ID)
+
+	// TODO: Handle the auth
+	deleteHTTPReq, err := http.NewRequest("DELETE", url, nil)
+	if err != nil {
+		log.Printf("Failed to create new HTTP request: %v", err)
+		return err
+	}
+
+	log.Printf("Doing a request to: %s", url)
+	resp, err := c.client.Do(deleteHTTPReq)
+	if err != nil {
+		log.Printf("Failed to DELETE: %#v", err)
+		return err
+	}
+	defer resp.Body.Close()
```

Comment:
@krancour does this still leak memory in Go 1.7+? I couldn't get it to leak on my host, so just wondering

(not blocking for this PR)

Refs:

- https://github.com/golang/go/issues/5314
- https://husobee.github.io/golang/memory/leak/2016/02/11/go-mem-leak.html